implementation/egads/src/main/java/net/opentsdb/query/anomaly/egads/olympicscoring/OlympicScoringNode.java

```diff
@@ -0,0 +1,815 @@
+    }
+    
+    // fire!
+    for (int i = 0; i < config.getBaselineNumPeriods(); i++) {
+      final BaselineQuery query = new BaselineQuery(i);
+      baseline_queries[i] = query;
+      query.sub_context = buildQuery((int) start.epoch(), 
+                                     (int) end.epoch(), 
+                                     context.queryContext(), 
+                                     query);
+      start.add(baseline_period);
+      end.add(baseline_period);
+    }
+
+    baseline_queries[0].sub_context.initialize(null)
+      .addCallback(new SubQueryCB(baseline_queries[0].sub_context))
+      .addErrback(new ErrorCB());
+    
+    if (config.getMode() == ExecutionMode.PREDICT) {
+      // return here.
+      final AnomalyPredictionState state = cache.getState(cache_key);
+      QueryExecutionException e = new QueryExecutionException("Successfully "
+          + "started prediction start [" + prediction_start + "] and key " 
+          + Arrays.toString(cache_key) + " State: " + JSON.serializeToString(state), 423);
+      sendUpstream(e);
+    }
+  }
+  
+  void countdown() {
+    latch.countDown();
+    if (latch.getCount() == 0) {
+      run();
+    }
+  }
+  
+  QueryContext buildQuery(final int start, 
+                          final int end, 
+                          final QueryContext context, 
+                          final QuerySink sink) {
+    final SemanticQuery.Builder builder = SemanticQuery.newBuilder()
+        // TODO - PADDING compute the padding
+        .setMode(QueryMode.SINGLE)
+        .setStart(Integer.toString(start - 300))
+        .setEnd(Integer.toString(end));
+    if (config.getBaselineQuery().getFilters() != null) {
+      builder.setFilters(config.getBaselineQuery().getFilters());
+    }
+    
+    for (final QueryNodeConfig config : config.getBaselineQuery().getExecutionGraph()) {
+      if (config instanceof DownsampleConfig) {
+        builder.addExecutionGraphNode(((DownsampleConfig.Builder)
+            config.toBuilder())
+            .setInterval(ds_interval)
+            .setSources(config.getSources())
+            .build());
+      } else {
+        builder.addExecutionGraphNode(config);
+      }
+    }
+    
+    LOG.info("  BASELINE Q: " + JSON.serializeToString(builder.build()));
+    return SemanticQueryContext.newBuilder()
+        .setTSDB(context.tsdb())
+        .setLocalSinks((List<QuerySink>) Lists.newArrayList(sink))
+        .setQuery(builder.build())
+        .setStats(context.stats())
+        .setAuthState(context.authState())
+        .setHeaders(context.headers())
+        .build();
+  }
+  
+  ChronoUnit modelDuration() {
+    return model_units;
+  }
+  
+  long predictionStart() {
+    return prediction_start;
+  }
+  
+  long predictionIntervals() {
+    return prediction_intervals;
+  }
+  
+  long predictionInterval() {
+    return prediction_interval;
+  }
+  
+  void writeCache(final QueryResult result) {
+    if (cache_hit || cache == null) {
```

Comment:
If `CacheCB.call()` is called with a non-`null` value, then it will not branch and will eventually set `prediction` to that argument, set `cache_hit` to `true`, and count down the latch. If it is possible for another thread to have entered `writeCache()` concurrently, then there is a data race on `cache_hit`. The `volatile` modifier itself does not resolve races.

There may be no problem here. I need more time to analyze program flow.

Reply:
Yeah I really just have volatile to make sure we don't cache the pointer. Only one path and one thread will ever call into this function. I just can't guarantee which thread that will be.